onert-micro/onert-micro/src/core/OMTrainingRuntimeModule.cpp

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/OMTrainingRuntimeModule.h"
+#include "import/OMExecutionPlanCreator.h"
+#include "train/OMBackpropExecute.h"
+#include "core/train/OMCheckpointSaver.h"
+#include "core/train/OMCheckpointLoader.h"
+
+using namespace onert_micro::core;
+using namespace onert_micro;
+
+/*
+ * Import train model, validate it, prepare necessary structures and data for future calculations
+ */
+OMStatus OMTrainingRuntimeModule::importTrainModel(char *model_ptr, const OMConfig &config)
+{
+  // 1 - Import main model
+  // 2 - Import and create training entities
+
+  // 1 - Import main model
+  OMStatus status = importModel(model_ptr, config);
+  if (status != Ok)
+    return status;
+
+  // 2 - Import and create training entities
+  _backward_graphs.resize(_graphs.size());
+  for (uint32_t i = 0; i < _backward_graphs.size(); ++i)
+  {
+    // load default backward graph
+    OMRuntimeGraph &graph = _backward_graphs.at(i);
+
+    OMRuntimeContext &runtime_context = graph.getRuntimeContext();
+    OMRuntimeStorage &runtime_storage = graph.getRuntimeStorage();
+    memory::OMRuntimeAllocator &runtime_allocator = graph.getRuntimeAllocator();
+
+    runtime_context.setModel(model_ptr, i);
+
+    // Parse and validate WOF file if it is exist
+    // WARNING: setWofFile method of RuntimeContext should follow after setModel.
+    if (config.wof_ptr != nullptr)
+      runtime_context.setWofFile(config.wof_ptr);
+
+    // AllocDeallocPlan backward graph creation
+    status = import::OMExecutionPlanCreator::createBackwardExecutionPlan(
+      runtime_storage, runtime_context, runtime_allocator, config);
+    if (status != Ok)
+      return status;
+  }
+
+  // Set current optimizer
+  _training_handler.setOptimizer(config);
+
+  return Ok;
+}
+
+/*
+ *  Train single step: run forward graph (with data which was set in SetInput) ->
+ *  -> calculate error (with target data which was set in SetTarget) ->
+ *  -> run backward graph -> update optimizer state -> after batch_size steps update weights
+ *  Warning: before using trainSingleStep call: 1) importTrainModel; 2) setInput; 3) setTarget
+ */
+OMStatus OMTrainingRuntimeModule::trainSingleStep(const OMConfig &config)
+{
```

Comment:
```suggestion
{
  config.training_context.num_step++;
```
we need this for update num_step for ADAM.

Reply:
Yes, and to save this value in checkpoints files (if needed method will call)

Reply:
I mean we need somewhere to keep this num_step value